File: backend/src/routes/placementDebug.ts
/* Copyright Contributors to the Open Cluster Management project */
import type { Http2ServerRequest, Http2ServerResponse, OutgoingHttpHeaders } from 'node:http2'
import { constants } from 'node:http2'
import type { RequestOptions } from 'node:https'
import { request } from 'node:https'
import { pipeline } from 'node:stream'
import { URL } from 'node:url'
import { getPlacementDebugAgent, getServiceAgent } from '../lib/agent'
import { logger } from '../lib/logger'
import { notFound, respond, respondInternalServerError } from '../lib/respond'
import { getAuthenticatedToken } from '../lib/token'

const proxyHeaders = [
  constants.HTTP2_HEADER_ACCEPT,
  constants.HTTP2_HEADER_ACCEPT_ENCODING,
  constants.HTTP2_HEADER_CONTENT_ENCODING,
  constants.HTTP2_HEADER_CONTENT_LENGTH,
  constants.HTTP2_HEADER_CONTENT_TYPE,
]
const proxyResponseHeaders = [
  constants.HTTP2_HEADER_CACHE_CONTROL,
  constants.HTTP2_HEADER_CONTENT_LENGTH,
  constants.HTTP2_HEADER_CONTENT_ENCODING,
  constants.HTTP2_HEADER_ETAG,
]

const defaultServiceHost = 'cluster-manager-placement.open-cluster-management-hub.svc.cluster.local'
const defaultPlacementDebugUrl = `https://${defaultServiceHost}:9443/debug/placements/`

const isProduction = process.env.NODE_ENV === 'production'

export async function placementDebug(req: Http2ServerRequest, res: Http2ServerResponse): Promise<void> {
  const token = await getAuthenticatedToken(req, res)
  if (!token) return

  const agent = isProduction ? getPlacementDebugAgent() : getServiceAgent()
  if (!agent) {
    return respond(res, { error: 'Placement debug service unavailable — OCM CA bundle not configured' }, 503)
  }

  const headers: OutgoingHttpHeaders = { authorization: `Bearer ${token}` }
  for (const header of proxyHeaders) {
    if (req.headers[header]) headers[header] = req.headers[header]
  }
  headers['content-type'] = 'application/json'

  const url = new URL(process.env.PLACEMENT_DEBUG_URL || defaultPlacementDebugUrl)
  headers.host = url.hostname

  const options: RequestOptions = {
    protocol: url.protocol,
    hostname: url.hostname,
    port: url.port,
    path: url.pathname,
    method: 'POST',
    headers,
    agent,
  }

  pipeline(
    req,
    request(options, (response) => {
      if (!response) return notFound(req, res)
      const responseHeaders: OutgoingHttpHeaders = { 'content-type': 'application/json' }
      for (const header of proxyResponseHeaders) {
        if (response.headers[header]) responseHeaders[header] = response.headers[header]
      }
      res.writeHead(response.statusCode ?? 500, responseHeaders)
      pipeline(response, res as unknown as NodeJS.WritableStream, (err) => {
        if (err) logger.error({ msg: 'placement debug response pipeline error', error: err.message })
      })
    }),
    (err) => {
      if (err) {
        logger.error({ msg: 'placement debug upstream error', error: err.message })
        if (!res.headersSent) respondInternalServerError(req, res)
      }
    }
  )
}

File: backend/internal/placementdebug/placementdebug.go
// Copyright Contributors to the Open Cluster Management project

package placementdebug

import (
	"crypto/tls"
	"encoding/json"
	"net/http"
	"net/http/httputil"
	"net/url"
	"os"
	"sync"
	"time"

	"k8s.io/client-go/rest"

	"github.com/stolostron/console/backend/internal/auth"
	"github.com/stolostron/console/backend/internal/outbound"
	applog "github.com/stolostron/console/backend/internal/log"
)

const defaultPlacementDebugURL = "https://cluster-manager-placement.open-cluster-management-hub.svc.cluster.local:9443/debug/placements/"

var requestHeaders = []string{
	"Accept",
	"Accept-Encoding",
	"Content-Encoding",
	"Content-Length",
	"Content-Type",
}

var responseHeaders = []string{
	"Cache-Control",
	"Content-Length",
	"Content-Encoding",
	"Etag",
}

// Options configure the placement-debug reverse proxy.
type Options struct {
	RESTConfig *rest.Config
	Authn      func(w http.ResponseWriter, r *http.Request) (string, bool)
	GetCA      func() []byte
	Endpoint   func() string
}

// Handler proxies POST /placement-debug.
type Handler struct {
	RESTConfig *rest.Config
	Authn      func(w http.ResponseWriter, r *http.Request) (string, bool)
	GetCA      func() []byte
	Endpoint   func() string

	mu    sync.Mutex
	proxy *httputil.ReverseProxy
	ca    string
}

// New returns a placement-debug proxy handler.
func New(opts Options) *Handler {
	h := &Handler{
		RESTConfig: opts.RESTConfig,
		Authn:      opts.Authn,
		GetCA:      opts.GetCA,
		Endpoint:   opts.Endpoint,
	}
	if h.Authn == nil && opts.RESTConfig != nil {
		h.Authn = func(w http.ResponseWriter, r *http.Request) (string, bool) {
			return auth.AuthenticateRequest(r.Context(), opts.RESTConfig, w, r)
		}
	}
	if h.Endpoint == nil {
		h.Endpoint = func() string {
			if v := os.Getenv("PLACEMENT_DEBUG_URL"); v != "" {
				return v
			}
			return defaultPlacementDebugURL
		}
	}
	return h
}

func (h *Handler) authenticate(w http.ResponseWriter, r *http.Request) (string, bool) {
	if h.Authn != nil {
		return h.Authn(w, r)
	}
	w.WriteHeader(http.StatusUnauthorized)
	return "", false
}

func (h *Handler) caPEM() []byte {
	if h.GetCA != nil {
		return h.GetCA()
	}
	return nil
}

func (h *Handler) reverseProxy(target *url.URL, ca []byte) *httputil.ReverseProxy {
	pem := string(ca)
	h.mu.Lock()
	defer h.mu.Unlock()
	if h.proxy != nil && h.ca == pem {
		return h.proxy
	}
	tlsCfg := &tls.Config{MinVersion: tls.VersionTLS12, RootCAs: nil}
	if len(ca) > 0 {
		tlsCfg = auth.TLSConfigFromCA(ca, false)
	}
	rp := &httputil.ReverseProxy{
		Rewrite: func(pr *httputil.ProxyRequest) {
			token := auth.TokenFromRequest(pr.In)
			pr.SetURL(target)
			pr.Out.URL.Path = target.Path
			pr.Out.URL.RawPath = target.RawPath
			pr.Out.URL.RawQuery = pr.In.URL.RawQuery
			pr.Out.Host = target.Hostname()
			pr.Out.Header = http.Header{}
			for _, name := range requestHeaders {
				if v := pr.In.Header.Get(name); v != "" {
					pr.Out.Header.Set(name, v)
				}
			}
			pr.Out.Header.Set("Content-Type", "application/json")
			pr.Out.Header.Set("Authorization", "Bearer "+token)
			pr.Out.Header.Set("Host", target.Hostname())
		},
		ModifyResponse: func(resp *http.Response) error {
			filtered := http.Header{"Content-Type": []string{"application/json"}}
			for _, name := range responseHeaders {
				for _, v := range resp.Header.Values(name) {
					filtered.Add(name, v)
				}
			}
			resp.Header = filtered
			return nil
		},
		ErrorHandler: func(w http.ResponseWriter, _ *http.Request, err error) {
			applog.Logger().Error("placement debug upstream error", "error", err)
			w.WriteHeader(http.StatusInternalServerError)
		},
		Transport: outbound.Transport(tlsCfg, false),
		FlushInterval: -1 * time.Millisecond,
	}
	h.proxy = rp
	h.ca = pem
	return rp
}

// ServeHTTP authenticates and reverse-proxies to the placement debug service.
func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusNotFound)
		return
	}
	if _, ok := h.authenticate(w, r); !ok {
		return
	}
	ca := h.caPEM()
	if len(ca) == 0 {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusServiceUnavailable)
		_ = json.NewEncoder(w).Encode(map[string]string{
			"error": "Placement debug service unavailable — OCM CA bundle not configured",
		})
		return
	}
	endpoint := ""
	if h.Endpoint != nil {
		endpoint = h.Endpoint()
	}
	target, err := url.Parse(endpoint)
	if err != nil || target.Scheme == "" || target.Host == "" {
		w.WriteHeader(http.StatusInternalServerError)
		return
	}
	h.reverseProxy(target, ca).ServeHTTP(w, r)
}
